src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -96,26 +166,137 @@ static void uct_cuda_ipc_cache_purge(uct_cuda_ipc_cache_t *cache)
     ucs_trace("%s: cuda ipc cache purged", cache->name);
 }
 
-static ucs_status_t uct_cuda_ipc_open_memhandle(const uct_cuda_ipc_rkey_t *key,
-                                                CUdeviceptr *mapped_addr)
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_legacy(CUipcMemHandle memh,
+                                   CUdeviceptr *mapped_addr)
 {
     CUresult cuerr;
-    ucs_status_t status;
 
-    cuerr = cuIpcOpenMemHandle(mapped_addr, key->ph,
+    cuerr = cuIpcOpenMemHandle(mapped_addr, memh,
                                CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS);
-    if (cuerr == CUDA_SUCCESS) {
-        status = UCS_OK;
-    } else {
+    if (cuerr != CUDA_SUCCESS) {
         ucs_debug("cuIpcOpenMemHandle() failed: %s",
                   uct_cuda_base_cu_get_error_string(cuerr));
-        status = (cuerr == CUDA_ERROR_ALREADY_MAPPED) ? UCS_ERR_ALREADY_EXISTS :
-                                                        UCS_ERR_INVALID_PARAM;
+        return (cuerr == CUDA_ERROR_ALREADY_MAPPED) ?
+            UCS_ERR_ALREADY_EXISTS : UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+#if HAVE_CUDA_FABRIC
+static ucs_status_t
+uct_cuda_ipc_open_memhandle_vmm(const uct_cuda_ipc_rkey_t *key,
+                                CUdeviceptr *mapped_addr)
+{
+    CUmemAccessDesc access_desc = {};
+    ucs_status_t status;
+    CUdeviceptr dptr;
+    CUmemGenericAllocationHandle handle;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemImportFromShareableHandle(&handle,
+                (void*)&key->ph.handle.vmm, CU_MEM_HANDLE_TYPE_FABRIC));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status =
+        UCT_CUDADRV_FUNC_LOG_ERR(cuMemAddressReserve(&dptr, key->b_len, 0, 0, 0));
+    if (status != UCS_OK) {
+        goto release_handle;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemMap(dptr, key->b_len, 0, handle, 0));
+    if (status != UCS_OK) {
+        goto release_va_range;
     }
 
+    cuCtxGetDevice(&access_desc.location.id);
+
+    access_desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    access_desc.flags         = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemSetAccess(dptr, key->b_len, &access_desc, 1));
+    if (status != UCS_OK) {
+        cuMemUnmap(dptr, key->b_len);
+        goto release_va_range;
+    }
+
+    *mapped_addr = dptr;
+
+    return UCT_CUDADRV_FUNC_LOG_ERR(cuMemRelease(handle));
```

Comment:
if cuMemRelease fails here, there is no cleanup

Reply:
Will fix this.